client/lib/src/server.rs

```diff
@@ -1,80 +1,70 @@
+#![allow(missing_docs)]
+
 //! A low-level interface to send and receive server-client protocol messages
 
 use std;
 
-use common::socket::{SendSocket, ReceiveSocket};
+use std::sync::{Arc, Mutex};
+use std::sync::mpsc::{Sender, Receiver, TryRecvError};
+use std::sync::atomic::{Ordering, AtomicUsize};
 
-#[allow(missing_docs)]
-pub mod send {
-  use std;
-  use std::sync::mpsc::Sender;
+use bincode;
 
-  use common::protocol;
+use common::protocol;
+use common::socket::{SendSocket, ReceiveSocket};
 
-  #[derive(Clone)]
-  pub struct T {
-    pub sender     : Sender<Vec<u8>>,
-    pub bytes_sent : std::sync::Arc<std::sync::Mutex<u64>>,
-  }
+#[derive(Clone)]
+pub struct SSender {
+  // A boxed slice is used to reduce the sent size
+  pub sender: Sender<Box<[u8]>>,
+  // Please replace with AtomicU64 when it becomes stable
+  pub bytes_sent: Arc<AtomicUsize>,
+}
 
-  pub fn new(sender: Sender<Vec<u8>>) -> T {
-    T {
-      sender     : sender,
-      bytes_sent : std::sync::Arc::new(std::sync::Mutex::new(0)),
+impl SSender {
+  pub fn new(sender: Sender<Box<[u8]>>) -> SSender {
+    SSender {
+      sender: sender,
+      bytes_sent: Arc::new(AtomicUsize::new(0)),
     }
   }
 
-  impl T {
-    pub fn tell(&self, msg: &protocol::ClientToServer) {
-      use bincode;
-      use bincode::serialize;
-      let msg = serialize(msg, bincode::Infinite).unwrap();
-      *self.bytes_sent.lock().unwrap() += msg.len() as u64;
-      self.sender.send(msg).unwrap();
-    }
+  pub fn tell(&self, msg: &protocol::ClientToServer) {
+    let msg = bincode::serialize(msg, bincode::Infinite).unwrap();
+    // We aren't reading this until long after the write, so we use `Relaxed`
+    self.bytes_sent.fetch_add(msg.len() as usize, Ordering::Relaxed);
+    self.sender.send(msg.into_boxed_slice()).unwrap();
   }
 }
 
-#[allow(missing_docs)]
-pub mod recv {
-  use bincode;
-  use std;
-  use std::sync::mpsc::Receiver;
-  use std::sync::mpsc::TryRecvError;
-
-  use common::protocol;
-
-  #[derive(Clone)]
-  pub struct T (pub std::sync::Arc<Receiver<Vec<u8>>>);
-
-  impl T {
-    pub fn try(&self) -> Option<protocol::ServerToClient> {
-      match self.0.try_recv() {
-        Ok(msg) => Some(bincode::deserialize(&msg).unwrap()),
-        Err(TryRecvError::Empty) => None,
-        e => {
-          e.unwrap();
-          unreachable!();
-        },
-      }
+#[derive(Clone)]
+pub struct SReceiver (Arc<Mutex<Receiver<Box<[u8]>>>>);
```

Comment:
Why is there an extra `Mutex` now?

Reply:
It has to be shared between threads, as `T` is both `Clone` and `Send`.

Reply:
Ahh I used to get around this (wrongly with the `unsafe impl`. I get it now thanks.